XiEditor/XiDocumentController.swift

```diff
@@ -105,83 +118,63 @@ class XiDocumentController: NSDocumentController {
+    /// Asks core to open the specified document. On success, we will receive
+    /// the `new_view` notification.
+    func openDocumentAsync(atUrl url: URL?) {
+        if let url = url, let existing = self.document(for: url) {
+            existing.showWindows()
+        } else {
+            let delegate = (NSApp.delegate as! AppDelegate)
+            let params = url == nil ? [:] : ["file_path": url!.path]
+            delegate.dispatcher?.coreConnection
+                .sendRpcAsync("new_view_async", params: params)
         }
-        fatalError("Xi does not currently support document duplication")
     }
 
-    func setupDocument(_ document: Document, forUrl url: URL?) {
-        // we nil out this field to opt out of having NSDocument check for changes on disk when saving;
-        // we (theoretically) do that check in xi-core.
-        document.fileModificationDate = nil
-        Events.NewView(path: url?.path).dispatchWithCallback(document.dispatcher!) { (response) in
-            DispatchQueue.main.sync {
+    func makeDocumentSync(atUrl url: URL?) throws -> NSDocument {
+        let newDocument = try makeUntitledDocument(ofType: defaultType!) as! Document
+        Events.NewView(path: url?.path).dispatchWithCallback(newDocument.dispatcher!) { (response) in
+            DispatchQueue.main.async {
                 switch response {
                 case .ok(let result):
                     let viewIdentifier = result as! String
-                    self.setIdentifier(viewIdentifier, forDocument: document)
-                    document.coreViewIdentifier = viewIdentifier
+                    self.lock.lock()
+                    self.openViews[viewIdentifier] = newDocument
+                    newDocument.coreViewIdentifier = viewIdentifier
+                    newDocument.fileURL = url
+                    newDocument.fileModificationDate = nil
+                    self.lock.unlock()
                 case .error(let error):
-                    document.close()
+                    newDocument.close()
                     (NSApplication.shared.delegate as! AppDelegate).alert(text: error.message)
                 }
             }
         }
+        return newDocument
+    }
+
+    // Called when we create a new view, or when a new view is created at startup.
+    override func openUntitledDocumentAndDisplay(_ displayDocument: Bool) throws -> NSDocument {
+        let document = try makeDocumentSync(atUrl: nil)
+        self.addDocument(document)
+        if displayDocument {
+            document.makeWindowControllers()
+            document.showWindows()
+        }
+        return document
+    }
+
+    // this is called when documents are restored on startup. We still use the
+    // sync API for creating a new view because we're still using NSDocumentController.
+    override func makeDocument(for urlOrNil: URL?, withContentsOf contentsURL: URL, ofType typeName: String) throws -> NSDocument {
```

Comment:
I feel like saying something is an optional is another way of saying that it can be nil - I think "urlOrNil" is really unnecessary.

Reply:
Yea, I agree with you. This method (on `NSDocumentController`) was introduced in the objc days, and objc naming conventions were more verbose. I wonder if perhaps the swift name was generated automatically by some conversion tool?